Update kernel commandline docs (SC-1457) (#2133)
canonical/cloud-init: doc/rtd/development/datasource_creation.rst

.. _datasource_creation:

Datasource creation
*******************

> Since there are multiple ways to provide `user data`, `meta data`, and

Reviewer: ```suggestion
There are multiple ways to provide `user data`, `meta data`, and
```

> `vendor data` (each cloud solution seems to prefer its own way), a datasource

Reviewer: ```suggestion
`vendor data`, and each cloud solution prefers its own way, a datasource
```

abstract class was internally created to allow for a single way to access the
different cloud systems methods, providing this data through the typical usage

> of subclasses. See :file:`cloud-init/sources/__init__.py` as a starting

Author: That sentence is a bunch of fluff and jargon.

How about:
```
An abstract base class defines a single interface to interact with the different clouds. Each cloud implementation must inherit from this base class to utilize this shared functionality and interface.```

> of subclasses. See :file:`cloud-init/sources/__init__.py` as a starting

Reviewer: Definitely better. I'd say "use" instead of "utilize", and I'd keep "datasource" with the "abstract base class" bit (so that we're being specific about which abstract base class we're talking about), but otherwise, yeah, much improved.

point for understanding the datastructures used.

> The datasource objects have a few touch points with ``cloud-init``. If you

Reviewer: Is a datasource object the same thing as the datasource class mentioned above?

Author: Good question.

There is a relationship between the [(abstract) parent class](https://github.com/canonical/cloud-init/blob/34637a49df13a054c4355de675b7cf286450b2f5/cloudinit/sources/__init__.py#L165) and the datasource object. The various child classes are defined in `cloudinit/sources/DataSource*.py` and inherit from `cloudinit.sources.DataSource`

As for this statement:

> The datasource objects have a few touch points with ``cloud-init``.

It does feel odd that we're trying to say that "there are several things that need to be updated when adding a new datasource", yet we use the word `object`, which only refers to only an instantance of the child class, not the creation of the child class itself. This wording seems odd to me, since adding a new datasource will always involve defining a new class, and is unlikely to require manipulating how the datasource object gets used.  Perhaps this would be more clear if it just said:

> Multiple parts of the codebase require an update when adding a new datasource.

Thoughts?

Reviewer: The reason why I asked is because it reads like we're introducing a new term or concept (in "database objects") that hasn't been defined in the page. I know objects usually have a specific meaning, so I don't think it hurts us to explicitly state that relationship you just described, unless (since it's a devdoc page) you think our readers will already understand all of that?

I do agree that stating "Multiple parts of the codebase require an update when adding a new datasource." is clearer than what was there originally, but should we also state explicitly which parts of the codebase? (or is that fully taken care of in the list that follows?)

are interested in adding a new datasource for your cloud platform you will
need to take care of the following items:

Identify a mechanism for positive identification of the platform
================================================================

It is good practice for a cloud platform to positively identify itself to
the guest. This allows the guest to make educated decisions based on the
platform on which it is running. On the x86 and arm64 architectures, many

> clouds identify themselves through DMI data. For example, Oracle's public

Author: It wouldn't hurt to link to something external. Not everyone will know.

That said, I'm not sure which link might be best:
[The standard would be most authoritative](https://www.dmtf.org/sites/default/files/standards/documents/DSP0005.pdf).
[Wikipedia might be most accessible](https://en.wikipedia.org/wiki/Desktop_Management_Interface)
[dmidecode may be most familiar to some](https://linux.die.net/man/8/dmidecode)

> clouds identify themselves through DMI data. For example, Oracle's public

Reviewer: I think probably the source is the best choice. If it were a public-facing doc, rather than a devdoc, I might be tempted to go with a more accessible option, but I think in this case it's better to link to the standard. It's the most likely to be kept up to date.

cloud provides the string ``'OracleCloud.com'`` in the DMI chassis-asset
field.

``Cloud-init``-enabled images produce a log file with details about the
platform. Reading through this log in :file:`/run/cloud-init/ds-identify.log`
may provide the information needed to uniquely identify the platform.
If the log is not present, you can generate it by running from source
:file:`./tools/ds-identify` or the installed location
:file:`/usr/lib/cloud-init/ds-identify`.

The mechanism used to identify the platform will be required for the
``ds-identify`` and datasource module sections below.

Add datasource module cloudinit/sources/DataSource<CloudPlatform>.py
====================================================================

> It is suggested that you start by copying one of the simpler datasources

Reviewer: ```suggestion
We suggest you start by copying one of the simpler datasources
```
